src/high-level/vector.lisp

```diff
@@ -0,0 +1,118 @@
+         (print-complex (z)
+           (format stream "~6,3f ~:[+~;-~]~6,3fj"
+                   (realpart z)
+                   (minusp (imagpart z))
+                   (abs (imagpart z)))))
+    (let* ((size (size vector))
+           (type (element-type vector))
+           (print-entry #'print-real)) ;; TODO: Check for complex type
+      (pprint-logical-block (stream nil)
+        (print-unreadable-object (vector stream :type t)
+          (format stream "~D:" size)
+          (dotimes (e size)
+            (pprint-newline :mandatory stream)
+            (funcall print-entry (tref vector e))))))))
+
+(set-pprint-dispatch 'vector 'pprint-vector)
+
+;;; Required abstract-tensor methods
+
+(defmethod rank ((vector vector))
+  (declare (ignore vector))
+  1)
+
+(defmethod shape ((vector vector))
+  (list (size vector)))
+
+(defmethod tref ((vector vector) &rest pos)
+  (assert (cl:= (rank vector) (list-length pos))
+          () "Invalid index ~a. Must be rank ~a" pos (rank vector))
+  (assert (cl:every #'< pos (shape vector))
+          () "Index ~a out of range" pos)
+  (aref (storage vector) (first pos)))
+
+(defmethod (setf tref) (new-value (vector vector) &rest pos)
+  (assert (cl:= (rank vector) (list-length pos))
+          () "Invalid index ~a. Must be rank ~a" pos (rank vector))
+  (assert (cl:every #'< pos (shape vector))
+          () "Index ~a out of range" pos)
+  (setf (aref (storage vector) (first pos)) new-value))
+
+(defmacro defvector (name type tensor-name matrix-name)
+  `(progn
+     (defclass ,name (vector)
+       ((storage :type (vector-storage ,type)))
+       (:documentation ,(format nil "Vector with element type of ~a" type)))
+     (defmethod update-instance-for-different-class :before
+         ((old ,tensor-name)
+          (new ,name)
+          &key)
+       (assert (cl:= 1 (rank old))))
+     (defmethod update-instance-for-different-class :before
+         ((old ,matrix-name)
+          (new ,name)
+          &key)
+       (assert (cl:= 1 (rank old))))
+     (defmethod update-instance-for-different-class :before
+         ((old ,name)
+          (new ,tensor-name)
+          &key)
+       (with-slots (shape rank) new
+         (setf shape (shape old)
+               rank 1)))
+     (defmethod update-instance-for-different-class :before
+         ((old ,name)
+          (new ,matrix-name)
+          &key)
+       (with-slots (shape rank) new
+         (setf shape (shape old)
+               rank 1)))))
+
+(defgeneric dot (vector1 vector2)
+  (:documentation "Compute the dot product of two vectors")
+  (:method ((vector1 vector) (vector2 vector))
+    (assert (cl:= (size vector1) (size vector2))
+            () "Vectors must have the same size. The first vector is size ~a and the second vector is size ~a."
+            (size vector1) (size vector2))
+    (loop :for i :below (size vector1)
+          :sum (* (tref vector1 i) (tref vector2 i)))))
+
+(defgeneric norm (vector)
```

Comment:
Yet another item for the wish list: support p-norms via an `&optional` argument `p`.

Reply:
Added in d6d2b7c